### PR TITLE
types: convert to new charset before inserting to blob or json column

### LINKED ISSUE #30690: The inserted result is different with blob type in gbk dml test
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
set names gbk;
drop table if exists t;
create table t (b blob, d json);
insert into t values ('你好', '{"测试": "你好"}');
select * from t;
```

### 2. What did you expect to see? (Required)
```
mysql> select * from t;
+----------------+----------------------+
| b              | d                    |
+----------------+----------------------+
| 0xE4BDA0E5A5BD | {"测试": "你好"}     |
+----------------+----------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
mysql> select * from t;
+----------------------+----------------------------+
| b                    | d                          |
+----------------------+----------------------------+
| 0xE6B5A3E78AB2E382BD | {"娴嬭瘯": "浣犲ソ"}       |
+----------------------+----------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

```
mysql> SELECT tidb_version();
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                            |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.4.0-alpha-411-g149bb438c-dirty
Edition: Community
Git Commit Hash: 149bb438c4f472fa1b4838859ee2cfffbd922d89
Git Branch: refactor-convert
UTC Build Time: 2021-12-13 10:02:50
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #30690

Problem Summary:

```sql
set names gbk;
drop table if exists t;
create table t (b blob, d json);
insert into t values ('你好', '{"测试": "你好"}');
select * from t;
```

The expected insertion procedure for TiDB is:

1. `'你好'` is parsed with GBK(because of the`@@character_set_client`), resulting a UTF-8 string `'浣犲ソ'` datum annotated with "gbk_bin"(because of the `@@character_set_connection`).
2. Before inserting to a **binary** type like blob and json, another conversion from UTF-8 to GBK is needed. Thus `'浣犲ソ'` should be changed back to `'你好'`.

Previously, step 2 is missed for type blob and json, this PR fix the issue. 
 
### What is changed and how it works?

This PR also moves string encoding/decoding/validation to `Datum.ConvertTo()`.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -1574,6 +1636,8 @@ func (d *Datum) convertToMysqlJSON(sc *stmtctx.StatementContext, target *FieldTy
 		}
 	case KindMysqlJSON:
 		ret = *d
+	case KindBinaryLiteral:
+		err = json.ErrInvalidJSONCharset.GenWithStackByArgs("binary")
```
Comment (Defined2014) [inline types/datum.go:1640]: Please add a test for this case

File: types/datum.go [hunk 2]
```diff
@@ -182,6 +184,51 @@ func (d *Datum) GetString() string {
...
+// GetStringWithCheck gets the string and checks if it is valid in a given charset.
+func (d *Datum) GetStringWithCheck(sc *stmtctx.StatementContext, chs string) (string, error) {
+	enc := charset.FindEncoding(chs)
+	if enc.Tp() == charset.EncodingTpUTF8 && sc.SkipUTF8Check ||
+		enc.Tp() == charset.EncodingTpASCII && sc.SkipASCIICheck {
+		return d.GetString(), nil
+	}
+	if chs == charset.CharsetUTF8 && !sc.SkipUTF8MB4Check {
+		enc = charset.EncodingUTF8MB3StrictImpl
+	}
```
Comment (Defined2014) [inline types/datum.go:222]: Could we add a func called `getEncoding` to remove duplicate code?

File: cmd/explaintest/r/new_character_set_invalid.result [hunk 3]
```diff
@@ -5,7 +5,7 @@ insert into t values ('中文', 'asdf', '字符集');
 insert into t values ('À', 'ø', '😂');
 Error 1366: Incorrect string value '\xC3\x80' for column 'a'
 insert into t values ('中文À中文', 'asdføfdsa', '字符集😂字符集');
-Error 1366: Incorrect string value '\xC3\x80\xE4\xB8\xAD\xE6...' for column 'a'
+Error 1366: Incorrect string value '\xC3\x80' for column 'a'
```
Comment (Defined2014) [inline cmd/explaintest/r/new_character_set_invalid.result:8]: It's not same as MySQL.

Comment (tangenta) [reply cmd/explaintest/r/new_character_set_invalid.result:8]: I think it is OK to have this difference?

File: types/datum.go (hunk 1 above)
Comment (tangenta) [reply types/datum.go:1640]: https://github.com/pingcap/tidb/pull/31031/files#diff-aa77b7b03663bb3d8eb4863157d94b2aa47f0ecec2e6876ea353eb825cadaf2fR219

Here it is :)

**Comment by tangenta** [comment]
`testColumnTypeChangeSuite.TestChangeFromBitToStringInvalidUtf8ErrMsg` is failed but I try it on MySQL 5.7.36 and 8.0.27. The result is different from the test:
```sql
mysql> create table t (a bit(45));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t values (117471723421);
Query OK, 1 row affected (0.01 sec)

mysql> alter table t modify column a varchar(31) collate utf8mb4_general_ci;
Query OK, 1 row affected (0.05 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> select * from t;
+--------------+
| a            |
+--------------+
| 117471723421 |
+--------------+
1 row in set (0.03 sec)
```

We can see that even if non-binary string, the `bit` type is changed to `int` first instead of `varchar` directly. This is conflict with https://github.com/pingcap/tidb/issues/25037#issuecomment-852985135:
> 2: when cast the bit to binary, we should consider to convert it to uint then cast uint to string, rather than takeing the bit to string directly.

File: cmd/explaintest/r/new_character_set_invalid.result (hunk 3 above)
Comment (Defined2014) [reply cmd/explaintest/r/new_character_set_invalid.result:8]: Em... Not sure. Seems we had a PR to fix it before. https://github.com/pingcap/tidb/pull/25087

Comment (tangenta) [reply cmd/explaintest/r/new_character_set_invalid.result:8]: This PR does not introduce the garbled code as described in #25087. It extracts the error message from `ErrInvalidCharacterSet`(which should be in a well format).

The only difference is the number of invalid bytes displayed.

File: types/datum.go [hunk 4]
```diff
@@ -950,7 +1004,17 @@ func (d *Datum) convertToString(sc *stmtctx.StatementContext, target *FieldType)
 	case KindFloat64:
 		s = strconv.FormatFloat(d.GetFloat64(), 'f', -1, 64)
 	case KindString, KindBytes:
-		s = d.GetString()
+		fromBinary := d.Collation() == charset.CollationBin
+		toBinary := target.Charset == charset.CollationBin
```
Comment (xiongjiwei) [inline types/datum.go:1008]: charset eq collation seems weird

File: types/datum.go [hunk 5]
```diff
@@ -1574,6 +1640,8 @@ func (d *Datum) convertToMysqlJSON(sc *stmtctx.StatementContext, target *FieldTy
 		}
 	case KindMysqlJSON:
 		ret = *d
+	case KindBinaryLiteral:
+		err = json.ErrInvalidJSONCharset.GenWithStackByArgs("binary")
```
Comment (xiongjiwei) [inline types/datum.go:1644]: binary could be constant?

**Review by Defined2014 — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Comment by tangenta** [comment]
/run-unit-test

```
[2021-12-30T07:50:53.565Z] FAIL: client_test.go:104: testRestoreClientSuite.TestIsOnline
[2021-12-30T07:50:53.565Z] 
[2021-12-30T07:50:53.565Z] client_test.go:105:
[2021-12-30T07:50:53.565Z]     c.Assert(s.mock.Start(), IsNil)
[2021-12-30T07:50:53.565Z] ... value *errors.withStack = listen tcp 0.0.0.0:41402: bind: address already in use ("listen tcp 0.0.0.0:41402: bind: address already in use")
```

**Comment by zimulala** [comment]
/run-unit-test

```

[2021-12-30T08:38:41.665Z] FAIL: db_test.go:40: testRestoreSchemaSuite.SetUpSuite
[2021-12-30T08:38:41.665Z] 
[2021-12-30T08:38:41.665Z] db_test.go:47:
[2021-12-30T08:38:41.665Z]     c.Assert(s.mock.Start(), IsNil)
[2021-12-30T08:38:41.665Z] ... value *errors.withStack = listen tcp 0.0.0.0:36262: bind: address already in use ("listen tcp 0.0.0.0:36262: bind: address already in use")
[2021-12-30T08:38:41.665Z] 
```

**Pull request merged**